mapiproxy/servers/default/asyncemsmdb/dcesrv_asyncemsmdb.c

```diff
@@ -0,0 +1,1229 @@
+		return;
+	}
+	ndr_print->depth = 1;
+	ndr_print->print = ndr_print_debug_helper;
+	ndr_print->no_newline = false;
+
+	ndr_print_mapistore_notification(ndr_print, "notification", &n);
+
+	OC_DEBUG(5, "Notification received for session: %s", p->emsmdb_session_str);
+
+	retval = mapistore_notification_subscription_get(mem_ctx, p->mstore_ctx, p->emsmdb_uuid, &r);
+	if (retval != MAPISTORE_SUCCESS) {
+		OC_DEBUG(0, "no subscription to process");
+		talloc_free(mem_ctx);
+		return;
+	}
+
+	OC_DEBUG(5, "%d subscriptions available:", r.v.v1.count);
+	ndr_print_mapistore_notification_subscription(ndr_print, "subscriptions", &r);
+	talloc_free(ndr_print);
+
+	/* Process notifications */
+	switch (n.v.v1.flags) {
+	case (sub_NewMail):
+		ret = process_newmail_notification(mem_ctx, p, &n, &r);
+		if (ret) {
+			OC_DEBUG(0, "[asyncemsmdb]: Failed to process newmail notification (error=0x%x)", ret);
+			talloc_free(mem_ctx);
+			return;
+		}
+		break;
+	default:
+		OC_DEBUG(0, "[asyncemsmdb]: Unsupported notification 0x%x", n.v.v1.flags);
+		talloc_free(mem_ctx);
+		return;
+	}
+	talloc_free(ndr_pull);
+	talloc_free(mem_ctx);
+
+	p->r->out.pulFlagsOut = talloc_zero(p->dce_call, uint32_t);
+	*p->r->out.pulFlagsOut = 0x1;
+
+	status = dcesrv_reply(p->dce_call);
+	if (!NT_STATUS_IS_OK(status)) {
+		OC_DEBUG(0, "EcDoAsyncWaitEx_handler: dcesrv_reply() failed - %s", nt_errstr(status));
+	}
+
+	talloc_free(p->fd_event);
+	p->fd_event = NULL;
+}
+
+
+ /**
+    \details exchange_async_emsmdb EcDoAsyncWaitEx (0x0) function
+
+   \param dce_call pointer to the session context
+   \param mem_ctx pointer to the memory context
+   \param r pointer to the EcDoAsyncWaitEx request data
+
+   \return NT_STATUS_OK on success
+ */
+static NTSTATUS dcesrv_EcDoAsyncWaitEx(struct dcesrv_call_state *dce_call,
+				       TALLOC_CTX *mem_ctx,
+				       struct EcDoAsyncWaitEx *r)
+{
+	enum mapistore_error			retval;
+	struct exchange_asyncemsmdb_session	*session = NULL;
+	struct asyncemsmdb_private_data		*p = NULL;
+	struct mapistore_context		*mstore_ctx = NULL;
+	struct GUID				uuid;
+	char					*cn = NULL;
+	const char				*ip_addr = NULL;
+	int					nn_retval;
+	size_t					sz = 0;
+	char					*bind_addr = NULL;
+	int					port = 0;
+
+	OC_DEBUG(3, "exchange_asyncemsmdb: EcDoAsyncWaitEx (0x0)");
+
+	/* Step 0. Ensure incoming user is authenticated */
+	if (!dcesrv_call_authenticated(dce_call)) {
+		OC_DEBUG(1, "No challenge requested by client, cannot authenticate");
+		*r->out.pulFlagsOut = 0x1;
+		return NT_STATUS_OK;
+	}
+
+	/* Set the asynchronous dcerpc flag on the connection */
+	dce_call->state_flags |= DCESRV_CALL_STATE_FLAG_ASYNC;
+
+	/* Step 1. Search for an existing session */
+	session = dcesrv_find_asyncemsmdb_session(&r->in.async_handle->uuid);
+	if (session) {
+		p = (struct asyncemsmdb_private_data *) session->data;
+		/* Ensure the session is registered */
+		retval = mapistore_notification_session_exist(p->mstore_ctx, r->in.async_handle->uuid);
```

Comment:
For me it would be more easy to read `session->emsmdb_uuid`. We have already fetched our session, so better use that instead of input parameters